libs/cardano-protocol-tpraos/src/Cardano/Protocol/TPraos/API.hs

```diff
@@ -516,3 +529,67 @@ getLeaderSchedule globals ss cds poolHash key pp = Set.filter isLeader epochSlot
     f = activeSlotCoeff globals
     epochSlots = Set.fromList [a .. b]
     (a, b) = runIdentity $ epochInfoRange ei currentEpoch
+
+class
+  ( PraosCrypto (Crypto era),
+    GetLedgerView era,
+    ApplyBlock era,
+    ApplyTx era,
+    CanStartFromGenesis era,
+    UsesValue era,
+    UsesScript era,
+    UsesAuxiliary era,
+    UsesTxBody era,
+    UsesTxOut era,
+    UsesPParams era,
+    ChainData (State (Core.EraRule "PPUP" era)),
+    SerialisableData (State (Core.EraRule "PPUP" era))
+  ) =>
+  ShelleyBasedEra era
+
+instance PraosCrypto crypto => ShelleyBasedEra (ShelleyEra crypto)
+
+mkInitialShelleyLedgerView ::
+  forall c.
+  ShelleyGenesis (ShelleyEra c) ->
+  LedgerView c
+mkInitialShelleyLedgerView genesisShelley =
+  LedgerView
+    { lvD = Shelley._d . sgProtocolParams $ genesisShelley,
+      lvExtraEntropy = Shelley._extraEntropy . sgProtocolParams $ genesisShelley,
+      lvPoolDistr = PoolDistr Map.empty,
+      lvGenDelegs = GenDelegs $ sgGenDelegs genesisShelley,
+      lvChainChecks = pparamsToChainChecksPParams . sgProtocolParams $ genesisShelley
+    }
+
+instance PraosCrypto c => ApplyTx (AllegraEra c)
```

Comment:
Unlike the shelley instance, I moved the other era instances here, and kept the `PraosCrypto` constraint. Do we like this better?